packages/next/build/webpack-config.ts

```diff
@@ -742,7 +745,9 @@ export default async function getBaseWebpackConfig(
       rules: [
         {
           test: /\.(tsx|ts|js|mjs|jsx)$/,
-          include: [dir, ...babelIncludeRegexes],
+          include: resolvedBaseUrl
+            ? [resolvedBaseUrl, dir, ...babelIncludeRegexes]
```

Comment:
The reason I haven't merged this is that this change needs to be thoroughly reviewed, I'm not entirely sure it's the correct behavior given that other packages in a repo are pretty much outside of the scope of Next.js and suddenly compiling them could be an issue, especially when deploying

Reply:
That's a valid concern:

> I'm not entirely sure it's the correct behavior given that other packages in a repo are pretty much outside of the scope of Next.js and suddenly compiling them could be an issue...

## Seems to require explicit opt-in

It is worth noting that by default nothing outside the Next root will be compiled - when you set up your project and start using typescript, the generated `tsconfig.json` does not include a `baseUrl`; so compilation will be limited to the Next folder.

A developer has to make a very explicit change to modify the `baseUrl`.

I'm trying to think why would anyone do this (modify baseUrl Next's `tsconfig`, either directly or through `extends`), if their intent is not to have code outside the Next folder compiled. But I cannot, for now at least, see this happening.

## Some background on mono-repos

I'm probably not the common Next user as I've been using Next nearly exclusively with mono-repos (either `docs` or `site`). And these have been either JS or TS.

You always do the same thing:

- You need to add the aliases to the webpack config where:
  - I always used my own bespoke alias resolver as you also often also need it for eslint and babel
  - then also config `baseUrl/paths`  in `tsconfig.json` to keep VScode happy
- Add `next-transpile-modules` so the other packages get compiled and bundled.

Our aliases would point to the `src` folder of the various packages, because Next would be there way before packages even get built into their own dist and published. Next builds for us are always apps/sites - a deployable rather than published package - you want everything bundled (at least everything in the mono-repo, some `node_modules` may flagged as externals, and loaded via `unpkg.com`) .

For us, we always want other packages bundled in both development and production builds - the whole point was that Next would build and bundle the source of other packages. It is an 'integration hub' if you wish.

Then there could be some extra effort in ensuring that when you build each package, aliases are not in force and one package does not bundle the other.

# What other use case then?

So I think there are two questions to be answered:

- When could a developer explicitly modify `baseUrl` yet not wish for anything under that `baseUrl` be complied?
- How could a developer mistakenly modify `baseUrl`?

Reply:
Although it's rather obvious, it is worth pointing out that with this change doesn't 2 things need to happen in order for stuff outside the Next project root to be compiled and bundled:

- The developer needs to explicitly set `baseUrl`.
- The developer has to import something from either outside the the Next root, or use an explicitly-set alias that points outside the Next root and use it.

So seems to me that there are already sufficient guards in place.

Reply:
I was more so targeting people that have already done that configuration and suddenly will have double-compiles, increased build times or broken code. Hence why this change in particular *might* need to be released separately. Ideally I'd like to land the other changes first anyway.

Reply:
It is a valid concern. But how all of this is possible?

This change simply tells the loader to also pick-up files that are under `baseUrl`. But the files must already be part of the dependency graph if they are ever to be compile - the loader won't be given anything that is not already in the graph. Right?

Reply:
Is the concern that people already modified the webpack config to build these files and now they are going to be built twice?

Reply:
Babel works in a way that it takes the input, parses it, and then applies transforms to that parsed input. The input would be the result of the previous loader, so the code itself would not change in most cases. It's not related to optimizations 👍

`.js` files could also be in the dependency graph without being passed through Babel if you just import them, e.g. `import MyLib from '../other-package/my-lib.js'`. This change would affect that if the root tsconfig already had a `baseUrl`

Reply:
OK. So what needs to be done to get all other changes merged?

Reply:
The `resolvedBaseUrl` should not be included and should be part of a different PR.

Reply:
So I've been busy lately but I did a quick rebase on origin/canary I haven't removed `resolvedBaseUrl` just yet

The benefits of the PR without it are quite limited....

Apart from my modification, `resolvedBaseUrl` and the `jsconfig.compilerOptions` which are affected by my changes because of the loading of the full typescript configuration are only used in https://github.com/vercel/next.js/blob/55e8bb46b9ac4f324c9a0b073980b8f5e1fa89a0/packages/next/build/webpack-config.ts#L987-L995

This means that webpack is going to resolve the paths using `tsconfig/jsconfig` paths but won't apply babel to the resolved module if they resolve to a path that's not below `dir` (aka the root of the module which uses nextjs).Therefore shared js typescript modules won't be transpiled. 
Users will still have to add or keep the `next-transpile-module` and ensure that the configuration of that plugin  stays in sync with the paths configuration they use.

I can't say I am 100% certain of my analysis though, this is all still quite new to me...

Reply:
I have updated the integration test to demonstrate this by using some typescript specific syntax in `lib` (the initial commit used typescript syntax which is compatible with esnext js syntax and would pass with or without the inclusion  of `resolvedBaseUr` in the transpilation configuration which wasn't very representative of what would happen in a real project)

Reply:
> This means that webpack is going to resolve the paths using tsconfig/jsconfig paths but won't apply babel to the resolved module if they resolve to a path that's not below dir (aka the root of the module which uses nextjs).Therefore shared js typescript modules won't be transpiled.

I'm aware that the changes without including resolvedBaseUrl are of limited use. The changes without it are fine to land and and optimization of existing features. Including the resolvedBaseUrl for compilation is a different story though and would require a full RFC to discuss with the community first given that it has more impact than you're considering right now. Could you write a RFC for changing the behavior.

Reply:
Alright, I'll see how to drop the change of `resolvedBaseUrl` out of the PR.
 
Be aware that none of the 3 issues (https://github.com/vercel/next.js/issues/15251, https://github.com/vercel/next.js/issues/13616, https://github.com/vercel/next.js/issues/9474) which mention this PR will be fixed by the PR alone.

Reply:
> Be aware that none of the 3 issues (#15251, #13616, #9474) which mention this PR will be fixed by the PR alone.

Yes I'm aware

Reply:
I have updated the PR to drop the change to babel loader's include pattern. the integration test demonstrates how one can manually configure the loader from `next.config.js` to get the behaviour back. 

>Could you write a RFC for changing the behavior.
I'll start an RFC topic as best I can.